// react-loaders-spinners/lib/components/BounceyLoader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
		value: true
});

var _templateObject = _taggedTemplateLiteral(['\n  0%, 75%, 100% {\n    transform: translateY(0px)\n  } \n  25% {\n    transform: translateY(-30px)\n  }\n'], ['\n  0%, 75%, 100% {\n    transform: translateY(0px)\n  } \n  25% {\n    transform: translateY(-30px)\n  }\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n\tdisplay: ', ';\n\talign-items: center;\n\n\tspan {\n\t\tmargin: ', ';\n\t\tdisplay: block;\n\t\twidth: ', ';\n\t\theight: ', ';\n\t\tborder-radius: 50%;\n\t\tbackground: ', ';\n\n\t\t&:nth-child(1) {\n\t\t\tanimation: ', ' 1s ease-in-out infinite;\n\t\t}\n\n\t\t&:nth-child(2) {\n\t\t\tanimation: ', ' 1s ease-in-out 0.33s infinite;\n\t\t}\n\n\t\t&:nth-child(3) {\n\t\t\tanimation: ', ' 1s ease-in-out 0.66s infinite;\n\t\t}\n\n\t\t&:nth-child(4) {\n\t\t\tanimation: ', ' 1s ease-in-out 0.66s infinite;\n\t\t}\n\t}\n'], ['\n\tdisplay: ', ';\n\talign-items: center;\n\n\tspan {\n\t\tmargin: ', ';\n\t\tdisplay: block;\n\t\twidth: ', ';\n\t\theight: ', ';\n\t\tborder-radius: 50%;\n\t\tbackground: ', ';\n\n\t\t&:nth-child(1) {\n\t\t\tanimation: ', ' 1s ease-in-out infinite;\n\t\t}\n\n\t\t&:nth-child(2) {\n\t\t\tanimation: ', ' 1s ease-in-out 0.33s infinite;\n\t\t}\n\n\t\t&:nth-child(3) {\n\t\t\tanimation: ', ' 1s ease-in-out 0.66s infinite;\n\t\t}\n\n\t\t&:nth-child(4) {\n\t\t\tanimation: ', ' 1s ease-in-out 0.66s infinite;\n\t\t}\n\t}\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _propTypes = require('prop-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var bounce = (0, _styledComponents.keyframes)(_templateObject);

var BounceLoader = _styledComponents2.default.div(_templateObject2, function (props) {
		return props.loading ? 'flex' : 'none';
}, function (props) {
		return props.spaceBetween ? '0 ' + props.spaceBetween + 'px' : '0 10px';
}, function (props) {
		return props.width ? props.width + 'px' : '20px';
}, function (props) {
		return props.height ? props.height + 'px' : '20px';
}, function (props) {
		return props.pColor ? props.pColor : '#555';
}, bounce, bounce, bounce, bounce);

var BounceyLoader = function BounceyLoader(props) {
		return _react2.default.createElement(
				BounceLoader,
				props,
				_react2.default.createElement('span', null),
				_react2.default.createElement('span', null),
				_react2.default.createElement('span', null)
		);
};

BounceyLoader.propTypes = {
		width: _propTypes.number,
		height: _propTypes.number,
		loading: _propTypes.bool.isRequired,
		pColor: _propTypes.string,
		spaceBetween: _propTypes.number
};

BounceyLoader.defaultProps = {
		width: 20,
		height: 20,
		pColor: '#555',
		loading: true,
		spaceBetween: 10
};

exports.default = BounceyLoader;
