// react-loaders-spinners/lib/components/ContentLoader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _templateObject = _taggedTemplateLiteral(['\n    from {\n      background-position: -', 'px 0;\n    }\n    to {\n      background-position: ', 'px 0;\n    }\n'], ['\n    from {\n      background-position: -', 'px 0;\n    }\n    to {\n      background-position: ', 'px 0;\n    }\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n  display: ', ';\n  flex-direction: column;\n  background: linear-gradient(to right, #eeeeee 8%, #e2e2e2 18%, #eeeeee 33%);\n  animation: ', ';\n  width: 100%;\n\n    div {\n      border: 8px solid #fff;\n      height: ', ';\n    }\n}\n'], ['\n  display: ', ';\n  flex-direction: column;\n  background: linear-gradient(to right, #eeeeee 8%, #e2e2e2 18%, #eeeeee 33%);\n  animation: ', ';\n  width: 100%;\n\n    div {\n      border: 8px solid #fff;\n      height: ', ';\n    }\n}\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _propTypes = require('prop-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var fadeLeftRight = function fadeLeftRight(props) {
  return (0, _styledComponents.keyframes)(_templateObject, props.width, props.width);
};

var Content = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.loading ? 'flex' : 'none';
}, function (props) {
  return fadeLeftRight(props) + ' 2s linear infinite';
}, function (props) {
  return props.lineHeight ? props.lineHeight + 'px' : '20px';
});

var divFactory = function divFactory(num) {
  var divs = [];
  for (var i = num; i > 0; i--) {
    divs.push(_react2.default.createElement('div', { key: i }));
  }
  return divs;
};

var ContentLoader = function ContentLoader(_ref) {
  var rows = _ref.rows,
      props = _objectWithoutProperties(_ref, ['rows']);

  return _react2.default.createElement(
    Content,
    props,
    divFactory(rows)
  );
};

ContentLoader.propTypes = {
  width: _propTypes.number.isRequired,
  loading: _propTypes.bool.isRequired,
  rows: _propTypes.number.isRequired,
  lineHeight: _propTypes.number
};

ContentLoader.defaultProps = {
  width: 200,
  rows: 4,
  loading: true,
  lineHeight: 20
};

exports.default = ContentLoader;
