// react-loaders-spinners/lib/components/SpinLoader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _templateObject = _taggedTemplateLiteral(['\n  0% {\n    transform: rotate(0deg);\n  }\n  100% {\n    transform: rotate(360deg);\n  }\n'], ['\n  0% {\n    transform: rotate(0deg);\n  }\n  100% {\n    transform: rotate(360deg);\n  }\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n  display: ', ';\n  border-radius: 50%;\n  border: ', ';\n  border-top: ', ';\n  height: ', ';\n  width: ', ';\n  animation: ', ' 1s linear infinite;\n'], ['\n  display: ', ';\n  border-radius: 50%;\n  border: ', ';\n  border-top: ', ';\n  height: ', ';\n  width: ', ';\n  animation: ', ' 1s linear infinite;\n']);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _propTypes = require('prop-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var spin = (0, _styledComponents.keyframes)(_templateObject);

var SpinLoader = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.loading ? 'flex' : 'none';
}, function (props) {
  return props.thickness && props.sColor ? props.thickness + 'px solid ' + props.sColor : props.thickness ? props.thickness + 'px solid #f3f3f3' : props.sColor ? '10px solid ' + props.sColor : '10px solid #f3f3f3';
}, function (props) {
  return props.thickness && props.pColor ? props.thickness + 'px solid ' + props.pColor : props.thickness ? props.thickness + 'px solid #555' : props.pColor ? '10px solid ' + props.pColor : '10px solid #555';
}, function (props) {
  return props.height ? props.height + 'px' : '50px';
}, function (props) {
  return props.width ? props.width + 'px' : '50px';
}, spin);

SpinLoader.propTypes = {
  width: _propTypes.number,
  height: _propTypes.number,
  thickness: _propTypes.number,
  loading: _propTypes.bool.isRequired,
  pColor: _propTypes.string,
  sColor: _propTypes.string
};

SpinLoader.defaultProps = {
  width: 50,
  height: 50,
  thickness: 10,
  sColor: '#f3f3f3',
  pColor: '#555',
  loading: true
};

exports.default = SpinLoader;
